torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -449,6 +482,98 @@ void ProcessGroupAgent::listenLoop() {
   }
 }
 
+void ProcessGroupAgent::pollTimedOutRPCs() {
+  while (!shutdown_.load()) {
+    std::chrono::milliseconds sleepTime;
+    std::unique_lock<std::mutex> lock{futureMutex_};
+    // Estimate amount of time the first future will time out in, and sleep
+    // for that long.
+    // if there are no futures or the first future's RPC timeout is set to 0
+    // (meaning no timeout), then sleep for a set "infinity" time.
+    if (futureTimeouts_.empty() ||
+        futureTimeouts_.begin()->first == INFINITE_TIMEOUT) {
+      sleepTime = INFINITE_TIMEOUT;
+    } else {
+      const auto minFutureExpirationTime = futureTimeouts_.begin()->first;
+      const auto remainingTime = getRPCRemainingTime(minFutureExpirationTime);
+      sleepTime = std::max(remainingTime, std::chrono::milliseconds(0));
+    }
+
+    if (sleepTime == INFINITE_TIMEOUT) {
+      futureTimeoutCV_.wait(lock);
```

Comment:
Should we add a `continue;` here? 

Since when being waken up from `::send(..)` call, the end times should not has arrived yet. There is no point of scanning the 2 maps?

Reply:
Doing this in https://github.com/pytorch/pytorch/pull/30355.